onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -2686,6 +2686,149 @@ It's an extension of Gelu. It takes the sum of input A and bias input B as the i
+      .Input(0, "input", "The input element.", "T")
+      .Output(0, "output", "The optional output enclosing the input element.", "O")
+      .TypeConstraint(
+          "T",
+          {"tensor(float)",
+           "seq(tensor(float))"},
+          "Constrains input type to all tensor and sequence types.")
+      .TypeConstraint(
+          "O",
+          {"optional(tensor(float))",
+           "optional(seq(tensor(float)))"},
+          "Constrains output type to all optional tensor or optional sequence types.")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          const size_t numInputs = ctx.getNumInputs();
+          if (numInputs != 1) {
+            fail_type_inference("OptionalConstruct is expected to have 1 input.");
+          }
+          auto input_type = ctx.getInputType(0);
+          if(input_type == nullptr){
+            fail_type_inference("Input type is null. Type info is expected.");
+          }
+          const size_t numOutputs = ctx.getNumOutputs();
+          if (numOutputs != 1) {
+            fail_type_inference("OptionalConstruct is expected to have 1 output.");
+          }
+          ctx.getOutputType(0)
+              ->mutable_optional_type()
+              ->mutable_elem_type()
+              ->CopyFrom(*input_type);
+        });
+
+  static const char* OptionalHasElement_ver1_doc = R"DOC(
+      Construct a optional type containing the 'input' element.
+      )DOC";
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(OptionalHasElement)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(OptionalHasElement_ver1_doc)
+      .Input(0, "input", "The optional input.", "O")
+      .Output(0, "output", "A boolean tensor of empty shape. If true, it indicates that optional input contains a tensor or sequence element. Otherwise, it is empty.", "B")
+      .TypeConstraint(
+          "O",
+          {"optional(tensor(float))",
+           "optional(seq(tensor(float)))"},
+          "Constrains input type to optional tensor and optional sequence types.")
+      .TypeConstraint(
+          "B",
+          {"tensor(bool)"},
+          "Constrains output to a boolean tensor.")
+      .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          const size_t numInputs = ctx.getNumInputs();
+          if (numInputs != 1) {
+            fail_type_inference("OptionalHasElement is expected to have 1 input.");
+          }
+          // const size_t numOutputs = ctx.getNumOutputs();
+          // if (numOutputs != 1) {
+          //   fail_type_inference("OptionalHasElement is expected to have 1 output.");
+          // }
+          // auto* output_tensor_type = ctx.getOutputType(0)->mutable_tensor_type();
+          // output_tensor_type->set_elem_type(TensorProto::BOOL);
+          // output_tensor_type->mutable_shape()->Clear();
+          });
+
+  static const char* OptionalGetElement_ver1_doc = R"DOC(
+      Outputs a copy of the element in the optional input'.
```

Comment:
We can drop "a copy of". I think that is an implementation detail. If the implementation can share the same tensor without copying, it can do so.

Reply:
Add "It is an error if the input value does not have an element and the behavior is undefined in this case."